File: Sem 6/DAA/Codes/NQueens.java
import java.util.*;

public class NQueens {
    private int[] x; // Array to store positions of queens
    private int n;  

    public NQueens(int n) {
        this.n = n;
        this.x = new int[n + 1]; 
    }

    public void solve() {
        nQueens(1);
    }

    private void nQueens(int k) {
        for (int i = 1; i <= n; i++) { // i is column
            if (place(k, i)) {
                x[k] = i;
                if (k == n) {
                    printSolution();
                } else {
                    nQueens(k + 1);
                }
            }
        }
    }

    private boolean place(int k, int i) {
        for (int j = 1; j < k; j++) { // j is row
            if (x[j] == i || Math.abs(x[j] - i) == Math.abs(j - k)) {
                return false;
            }
        }
        return true;
    }

    private void printSolution() {
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (x[i] == j) {
                    System.out.print("Q ");
                } else {
                    System.out.print(". ");
                }
            }
            System.out.println();
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int n = 4; // Example for 8 queens problem
        NQueens queens = new NQueens(n);
        queens.solve();
    }
}

File: sem6/DAA/Codes/nQueens.py
class nQueens:
    def __init__ (self, n):
        self.n = n
        self.x = [0] * (n+1)

    def solve(self):
        self.queens(1)

    def queens(self, k):
        for i in range(1, self.n + 1):
            if self.place(k, i):
                self.x[k] = i
                if k == self.n:
                    self.print_solution()
                else:
                    self.queens(k + 1)

    def place(self, k, i):
        for j in range(1, k):
            if self.x[j] == i or abs(self.x[j] - i) == abs(j-k):
                return False
        return True

    def print_solution(self):
        for i in range(1, self.n + 1):
            row = ['.'] * self.n
            row[self.x[i] - 1] = 'Q'
            print(' '.join(row))
        print()

if __name__ == "__main__":
    n = int(input("enter 'n' for dimension of chess board: "))
    queens = nQueens(n)
    queens.solve()
